nomad/node_endpoint.go

```diff
@@ -245,12 +245,32 @@ func (n *Node) constructNodeServerInfoResponse(snap *state.StateSnapshot, reply
 	return nil
 }
 
-// Deregister is used to remove a client from the cluster. If a client should
-// just be made unavailable for scheduling, a status update is preferred.
+// Deregister removes a single client node from the cluster
 func (n *Node) Deregister(args *structs.NodeDeregisterRequest, reply *structs.NodeUpdateResponse) error {
+	// Forward the request using the singular endpoint for backwards compatibility in
+	// clusters containing pre-0.9.4 servers
 	if done, err := n.srv.forward("Node.Deregister", args, args, reply); done {
 		return err
 	}
+
+	// If we're handling the request locally on the leader, operate on a batch of size 1
+	return n.batchDeregister(&structs.NodeBatchDeregisterRequest{
```

Comment:
:( I don't think this is right yet. Even if we are the leader, it's not safe to use the new raft message type until we confirm that all servers are at least 0.9.4. I would suggest keeping old handling untouched without it depending on batchDeregister (module some function extraction as preferred).